### ISSUE TITLE
JFR based meter registry

### ISSUE THREAD

**Issue by marschall** [micrometer-metrics/micrometer#2273]
We are currently using JDK Flight Recorder for our monitoring and metric needs. Recently a third party dependency introduced a dependency to Micrometer into our application. Following this we started implementing a meter registry that generates JFR events https://github.com/marschall/micrometer-jfr. We would be willing to contribute this if there is interest for it.

The implementation tries to do things the "JFR way":

- Uses JFR periodic events for sampled meters.
- Tries to add as much JFR metadata as possible for nice presentation in JDK Mission Control.
- Generates a JFR event whenever possible and lets statistic generation and histograms be done by whoever reads the recording. It therefore takes various shortcuts when it comes to statistics and histograms that I'm not sure are in the spirit of Micrometer. I have to confess I'm not very familiar with Micrometer concepts apart from skimming the docs.

JFR tequires an OpenJDK 11+, OpenJ9 won't work, Zing may work. There are some vendors that backported JFR to JDK 8, that may work.

**Comment by marschall** [comment]
Is this something you would be interested in? Should I start working on a PR?

**Comment by wasylni** [comment]
It would be great to see micrometer being able to consume/expose JFR events in Java 8/11.

**Comment by checketts** [comment]
@wasylni  Could you please add more details to your usecase? JFR sounds interesting. Which JFR events would be consumable and how might we use those in Micrometer?

**Comment by shakuzen** [comment]
@marschall This sounds interesting. If there's enough interest from others to use this, we can consider adding it here as an officially supported registry. There's nothing wrong with it being a third-party registry for the time being, though. I'm personally not very familiar with JFR for general purpose monitoring/metrics as opposed to profiling.

@wasylni As far as I understand, the proposal here does not deal with consuming JFR events, but rather with generating JFR events for corresponding Micrometer metrics. I suppose the reverse could be interesting as well, but if that's not part of the scope of this issue, maybe a separate issue and description of your use case would be best.

**Comment by marschall** [comment]
@shakuzen yes, the proposal in this issue is for a meter registry generating JFR events for corresponding Micrometer metrics
